MLP on permutedMNIST

karangrewal opened the pull request:
Here, I do the following:

1. add configuration files for training a 3- and 10-layer MLP on permutedMNIST, and
2. add an MLP class with modified weight initialization used by the configs to `nupic/research/frameworks/pytorch/models` so future MLP experiments can also use these.

benja-matic reviewed (commented):
Overall implementation looks good. Couple things I’d prefer to change though.

First, I think the experiments directory should contain experiments only, and not the class defitions for models. I suggest moving the ContinualLearningMLP class elsewhere (see below).

Second, the name ContinualLearningMLP is confusing since nothing in the class definition relates to continual learning in a way that’s obvious to me. As far as I can tell, this class just adds Kaiming initialization while taking into account input, weight, and activation sparsity – a great contribution by the way. But I think this belongs in either …frameworks/pytroch/models, or …frameworks/vernon/mixins, and could be named something like SparseKaimingInit, or something more specific to the code in the class def. Seems like a great piece of code to have available to anyone using an MLP, hence the proposal to move it to where others can see and use it.

Finally, I’m just curious where the learning rates come from. I vaguely remember you said you did some hyperparameter search in order to present the strongest results from the plain MLPs. If so, either the code for those experiments or at least some documentation / comments would be helpful.

karangrewal commented:
@benja-matic This is good feedback. To address your concerns,

1. I moved the MLP class to `nupic/research/frameworks/pytorch/models` and renamed it `ModifiedInitStandardMLP` and included some documentation there, and
2. I added comments regarding hyperparameter choices so they don't seem arbitrary.

benja-matic reviewed (approved):
Awesome, thanks for the adjustment. Looks great. You're missing an underscore online 28 in common_models but other than that, good to merge.
